ORC-1610: Reduce the number of hash computation in `CuckooSetBytes`

cxzl25 opened the pull request:
### What changes were proposed in this pull request?
Add boundary conditions on "length" with the min/max length stored in the hashes. 

### Why are the changes needed?
https://issues.apache.org/jira/browse/HIVE-24205

> This would significantly reduce the number of hash computation that needs to happen.

```
main insert:00:00:00.689
main lookup:00:00:01.124
PR insert:00:00:00.628
PR lookup:00:00:01.055
```

```java
  @Test
  public void testLen() {
    int maxSize = 200000;
    Random gen = new Random();
    String[] strings = new String[maxSize];
    for (int i = 0; i < maxSize; i++) {
      strings[i] = RandomStringUtils.random(Math.abs(gen.nextInt(1000)));
    }
    byte[][] values = getByteArrays(strings);

    StopWatch mainSW = new StopWatch();

    // load set
    mainSW.start();
    CuckooSetBytes main = new CuckooSetBytes(strings.length);
    main.fastLookup = false;
    for (byte[] v : values) {
      main.insert(v);
    }
    mainSW.split();
    System.out.println("main insert:" + mainSW);
    // test that the values we added are there
    for (byte[] v : values) {
      assertTrue(main.lookup(v, 0, v.length));
    }
    mainSW.stop();
    System.out.println("main lookup:" + mainSW);

    StopWatch prSW = new StopWatch();
    prSW.start();
    CuckooSetBytes pr = new CuckooSetBytes(strings.length);
    pr.fastLookup = true;
    for (byte[] v : values) {
      pr.insert(v);
    }
    prSW.split();
    System.out.println("PR insert:" + prSW);
    for (byte[] v : values) {
      assertTrue(pr.lookup(v, 0, v.length));
    }
    prSW.stop();
    System.out.println("PR lookup:" + prSW);
  }
```
### How was this patch tested?
GA

### Was this patch authored or co-authored using generative AI tooling?
No

dongjoon-hyun reviewed (commented):
Thank you for making a PR.

dongjoon-hyun reviewed (commented):
BTW, there is no official Apache Hive release with this patch, isn't it, @cxzl25 ?

dongjoon-hyun commented on java/core/src/java/org/apache/orc/util/CuckooSetBytes.java:
```diff
@@ -84,6 +86,9 @@ public CuckooSetBytes(int expectedSize) {
    * and ending at start+len is present in the set.
    */
   public boolean lookup(byte[] b, int start, int len) {
+    if (len < minLen || len > maxLen) {
```
In this class, `lookup` seems to be the first method when we call `insert`. Where `maxLen` is initialized before this usage?

cxzl25 replied:
`maxLen` is declared in the class member variable, so its initial value is 0, and the initial value of `minLen` is `Integer.MAX_VALUE`, so the first element found here basically returns false directly, and then calls insert, which When `minLen` and `maxlen` will be reassigned.

cxzl25 commented:
> BTW, there is no official Apache Hive release with this patch

[release-4.0.0-alpha-1](https://github.com/apache/hive/releases/tag/rel%2Frelease-4.0.0-alpha-1)
 on Mar 30, 2022

dongjoon-hyun commented:
No, that's 'alpha`. Nobody uses it seriously, @cxzl25 .

dongjoon-hyun reviewed (commented):
I'm wondering how you did verify the claim's PR? Did you see the actual perf benefits in a way, @cxzl25 ?

cxzl25 commented:
> No, that's 'alpha`. Nobody uses it seriously.

Thanks for your comment, let me close this PR first.

> I'm wondering how you did verify the claim's PR? Did you see the actual perf benefits in a way

I wrote a simple test that generated a random string and tested insert and lookup.

```
main insert:00:00:00.689
main lookup:00:00:01.124
PR insert:00:00:00.628
PR lookup:00:00:01.055
```

```java
  @Test
  public void testLen() {
    int maxSize = 200000;
    Random gen = new Random();
    String[] strings = new String[maxSize];
    for (int i = 0; i < maxSize; i++) {
      strings[i] = RandomStringUtils.random(Math.abs(gen.nextInt(1000)));
    }
    byte[][] values = getByteArrays(strings);

    StopWatch mainSW = new StopWatch();

    // load set
    mainSW.start();
    CuckooSetBytes main = new CuckooSetBytes(strings.length);
    main.fastLookup = false;
    for (byte[] v : values) {
      main.insert(v);
    }
    mainSW.split();
    System.out.println("main insert:" + mainSW);
    // test that the values we added are there
    for (byte[] v : values) {
      assertTrue(main.lookup(v, 0, v.length));
    }
    mainSW.stop();
    System.out.println("main lookup:" + mainSW);

    StopWatch prSW = new StopWatch();
    prSW.start();
    CuckooSetBytes pr = new CuckooSetBytes(strings.length);
    pr.fastLookup = true;
    for (byte[] v : values) {
      pr.insert(v);
    }
    prSW.split();
    System.out.println("PR insert:" + prSW);
    for (byte[] v : values) {
      assertTrue(pr.lookup(v, 0, v.length));
    }
    prSW.stop();
    System.out.println("PR lookup:" + prSW);
  }
```

dongjoon-hyun commented:
Thank you, @cxzl25 . Please add the above code and result into the PR description.

dongjoon-hyun commented:
Merged to main/2.0 for Apache ORC 2.0.0. Thank you again!
